lib/std/sort/pdqsort.zig

```diff
@@ -0,0 +1,564 @@
+
+    // limit the number of imbalanced partitions
+    const limit = math.floorPowerOfTwo(usize, items.len) + 1;
+
+    recurse(T, items, null, limit, context, lessThan);
+}
+
+/// sorts `orig_items` recursively.
+///
+/// If the slice had a predecessor in the original array, it is specified as
+/// `orig_pred` (must be the minimum value if exist).
+///
+/// `orig_limit` is the number of allowed imbalanced partitions before switching to `heapsort`.
+/// If zero, this function will immediately switch to heapsort.
+fn recurse(
+    comptime T: type,
+    orig_items: []T,
+    orig_pred: ?T,
+    orig_limit: usize,
+    context: anytype,
+    comptime lessThan: fn (context: @TypeOf(context), lhs: T, rhs: T) bool,
+) void {
+    var items = orig_items;
+    var pred = orig_pred;
+    var limit = orig_limit;
+
+    // slices of up to this length get sorted using insertion sort.
+    const max_insertion = 24;
+
+    // True if the last partitioning was reasonably balanced.
+    var was_balanced = true;
+
+    // True if the last partitioning didn't shuffle elements (the slice was already partitioned).
+    var was_partitioned = true;
+
+    while (true) {
+        // Very short slices get sorted using insertion sort.
+        if (items.len <= max_insertion) {
+            sort.insertionSort(T, items, context, lessThan);
+            return;
+        }
+
+        // If too many bad pivot choices were made, simply fall back to heapsort in order to
+        // guarantee `O(n log n)` worst-case.
+        if (limit == 0) {
+            heapify(T, items, context, lessThan);
+            return;
+        }
+
+        // If the last partitioning was imbalanced, try breaking patterns in the slice by shuffling
+        // some elements around. Hopefully we'll choose a better pivot this time.
+        if (!was_balanced) {
+            breakPatterns(T, items);
+            limit -= 1;
+        }
+
+        // Choose a pivot and try guessing whether the slice is already sorted.
+        var likely_sorted = false;
+        const pivot_idx = chosePivot(T, items, &likely_sorted, context, lessThan);
+
+        // If the last partitioning was decently balanced and didn't shuffle elements, and if pivot
+        // selection predicts the slice is likely already sorted...
+        if (was_balanced and was_partitioned and likely_sorted) {
+            // Try identifying several out-of-order elements and shifting them to correct
+            // positions. If the slice ends up being completely sorted, we're done.
+            if (partialInsertionSort(T, items, context, lessThan)) return;
+        }
+
+        // If the chosen pivot is equal to the predecessor, then it's the smallest element in the
+        // slice. Partition the slice into elements equal to and elements greater than the pivot.
+        // This case is usually hit when the slice contains many duplicate elements.
+        if (pred != null and pred.? == items[pivot_idx]) {
```

Comment:
When I was using this sort for my benchmark, I hit a compile error here, because my type was an aggregate and didn't have `==`.

Given that we pass `lessThan` as a parameter, perhaps we should pass `.eql` or the whole `compareFn` here? Or I guess using `std.meta.eql` would also mostly _just work_ here?

Reply:
I think a function returning std.math.Order would be most idiomatic

Reply:
One potential counter-argument here is that `compareFn` is not necessary the most optimal. Like if the items are strings, `==` can short circuit by comparing lengths, without doing extra work to figure out which one is actually smaller/larger. 

But I would be surprised if there is a case where this actually hits us, so, yeah, going with `compareFn` until someone presents a specific benchmark demonstrating unoptimality is a safe default choice.

Reply:
maybe a struct which implements something like?
```zig
const Sortable = struct {
    // if more specific functions are not implemented, this one is used
    pub fn cmp(_: anytype, lhs: T, rhs: T) std.math.Order {}
    
    // otherwise, these are optional:
    pub fn eql(ctx: anytype, lhs: T, rhs: T) bool {
        return cmp(ctx, lhs, rhs) == .eq;
    }
    pub fn lessThan(_: anytype, lhs: T, rhs: T) bool {
        return cmp(ctx, lhs, rhs) == .lt;
    }
};
```

Reply:
i went with @ifreund solution since it works for all cases and doesn't break anything if you already using `sort.asc` and `sort.desc`